### ISSUE TITLE
First sync takes days too complete both on testnet and public chain

### ISSUE THREAD

**Issue by kowalski** [ethereum/go-ethereum#2971]
#### System information

Geth version: Geth/v1.4.10-stable/linux/go1.5.1
Hardware: AWS instance of t2.small type (2 GB of memory)
```
$ cat /proc/cpuinfo 
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 63
model name	: Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz
stepping	: 2
microcode	: 0x25
cpu MHz		: 2400.072
cache size	: 30720 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm xsaveopt fsgsbase bmi1 avx2 smep bmi2 erms invpcid
bogomips	: 4800.14
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```
Open incoming ports: 30303 TCP and UDP
Command line used to run geth: 
```
/usr/bin/geth --fast --cache=128 --ipcpath /root/.ethereum/geth.ipc --rpc --rpcapi "eth,web3" --datadir /geth/data --password /geth/geth.password --rpcaddr "0.0.0.0" --rpcport 8110 
```
#### Expected behaviour

I'm running geth client on client data directory. I expect it to get synced in a reasonable time. I tried both with and without `--testnet`

#### Actual behaviour

The sync goes fast and than it becomes extremely slow. 
On testnet I have seen it go very fast to the block ~1.2M (with ~1.5M as head). On public net it went to ~1M (with ~2.2M as head). Actually I think I'm facing two separates problems.

#### Testnet

On testnet I can see I'm still going forward, but the speed it proceeds makes me sad. I'm doing roughly 2 blocks/s. 
```
> eth.syncing
{
  currentBlock: 1506000,
  highestBlock: 1567910,
  knownStates: 0,
  pulledStates: 0,
  startingBlock: 1461690
}
> net.peerCount
24
```
So, it seems it will complete in ~12 hours which will make it a total of 4 days to download the chain. 

#### Public net
Here as I mentioned before, the sync was very fast until the level of ~1M. After few hours it became really slow and now it's stuck completely:

```
{
  currentBlock: 1235819,
  highestBlock: 2172210,
  knownStates: 7418,
  pulledStates: 1041,
  startingBlock: 1235796
}
> net.peerCount
7
```

It doesn't move from here and will likely never finish.
Below is tail of my log output:

```
$ docker-compose -f docker-compose-staging.yml logs --tail 10 -f geth_public 
Attaching to admin_geth_public_1
geth_public_1          | I0831 08:24:06.388282 eth/handler.go:295] Peer bf63aae5026da2e6 [eth/63]: timed out DAO fork-check, dropping
geth_public_1          | I0831 08:24:06.498048 eth/downloader/downloader.go:274] Unregister failed: peer is not registered
geth_public_1          | I0831 08:24:38.644946 eth/downloader/downloader.go:274] Unregister failed: peer is not registered
geth_public_1          | I0831 08:24:58.095294 eth/handler.go:295] Peer bf63aae5026da2e6 [eth/63]: timed out DAO fork-check, dropping
geth_public_1          | I0831 08:24:58.095445 eth/downloader/downloader.go:274] Unregister failed: peer is not registered
geth_public_1          | I0831 08:25:49.934401 eth/downloader/downloader.go:1437] imported 384 state entries in 2m18.780812796s: processed 1041, pending at least 6336
geth_public_1          | I0831 08:25:49.940736 eth/handler.go:205] Removal failed: peer is not registered
geth_public_1          | I0831 08:25:49.940820 eth/handler.go:205] Removal failed: peer is not registered
geth_public_1          | I0831 08:25:49.941125 eth/handler.go:205] Removal failed: peer is not registered
geth_public_1          | I0831 08:25:50.351681 core/blockchain.go:751] imported 5 receipt(s) (0 ignored) in 414.177851ms. #1235819 [2ab1b91e… / d6f8f1aa…]

```

### This I've already tried with no avail

* I read about problems with possible system clock delay. It's not the case for me:
```
$ sudo ntpdate -v time.nist.gov
31 Aug 08:30:36 ntpdate[23836]: ntpdate 4.2.6p5@1.2349-o Fri Jul 22 17:30:52 UTC 2016 (1)
31 Aug 08:30:43 ntpdate[23836]: adjust time server 132.163.4.103 offset -0.030785 sec
```

* I tried different `--cache=..` settings (up to `1024`) and it doesn't seem to make the download visible faster, it does make it use much more memory though

**Comment by fjl** [comment]
Please try [geth 1.4.11](https://github.com/ethereum/go-ethereum/releases/tag/v1.4.11), it fixes a bunch of sync issues.

**Comment by kowalski** [comment]
Thank you Felix for looking into it. I've updated both public and testnet nodes to `Geth/v1.4.11-stable/linux/go1.5.1`. Here my observations:

### public chain

It's still slow, although `timed out DAO fork-check, dropping` messages are gone from the log output. 
However a new worrying one appeared: ` Synchronisation failed: block download canceled (requested)`.
More context:

```
Attaching to admin_geth_public_1
geth_public_1          | I0831 11:20:36.781997 eth/downloader/downloader.go:298] Synchronisation failed: block download canceled (requested)
geth_public_1          | I0831 11:21:14.387552 eth/downloader/downloader.go:605] Peer 2e85395c879b4f3e [hs 0.00/s, bs 0.00/s, rs 0.00/s, ss 0.00/s, miss    0, rtt 20s]: empty head header set
geth_public_1          | I0831 11:21:30.052117 eth/downloader/downloader.go:298] Synchronisation failed: block download canceled (requested)
geth_public_1          | I0831 11:21:37.151349 eth/downloader/downloader.go:948] imported 1 state entries in 36.69753ms: processed 1, pending at least 17
geth_public_1          | I0831 11:21:37.366039 eth/downloader/downloader.go:948] imported 2 state entries in 64.033378ms: processed 3, pending at least 49
geth_public_1          | I0831 11:21:40.469555 eth/downloader/downloader.go:948] imported 12 state entries in 174.761215ms: processed 15, pending at least 241
geth_public_1          | I0831 11:21:41.026988 eth/downloader/downloader.go:948] imported 32 state entries in 626.957347ms: processed 47, pending at least 753
geth_public_1          | I0831 11:21:47.806743 eth/downloader/downloader.go:948] imported 384 state entries in 3.086651229s: processed 431, pending at least 6762
geth_public_1          | I0831 11:22:06.802551 eth/downloader/downloader.go:948] imported 32 state entries in 18.157473821s: processed 463, pending at least 7274
geth_public_1          | I0831 11:22:07.200247 eth/downloader/downloader.go:298] Synchronisation failed: header processing canceled (requested)
geth_public_1          | I0831 11:22:58.089912 eth/handler.go:290] Peer 991aa11d7aa1cc4c [eth/63]: timed out DAO fork-check, dropping
```
What I truely find bizare is I get the following output on the console:
```
> eth.syncing
{
  currentBlock: 1247132,
  highestBlock: 1604998,
  knownStates: 10951,
  pulledStates: 463,
  startingBlock: 1247132
}
> net.peerCount
10
```
Previously the `highestBlock` was around `2.17M`. I don't exactly understand what happened with 500k missing blocks. Maybe this is not a problem, perhaps the first peer I connected too is also not yet fully synchronized.
I'm actually considering removing the chain and starting synchronization from scratch. What do you think ?

### testnet chain

It's still proceeding steadly at about 2 blocks/s. I can see no worrying messages in the console. So it seems to work, it's just very slow.

**Comment by kowalski** [comment]
Update: I've been investigating the issue and it seems my machine gets ~80% CPU wait. I've the geth nodes configured to keep their data dir on EBS volumes. The default EBS drive has the IOPS limit of 100. Now that I look at iostat I can see the following:
```
$ iostat -d -c -x
Linux 3.16.0-4-amd64 (ip-172-31-13-3) 	08/31/2016 	_x86_64_	(1 CPU)

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
          12.76    0.00    1.18   80.58    3.39    2.09

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
...
xvdg              0.50    11.51   64.71   18.22  1862.91  1903.06    90.83     6.84   82.44   33.61  255.88  11.21  92.95
```
(I removed the drives which are not essential)

The `xvdg` is the drive I used as a datadir for testnet, the one which proceeds the sync at the steady pace of about 2 blocks/second. `iostat` shows that at this pace it already uses `82 IOPS (64r + 18w)`. 

Summing up, I think the issue is in a way `geth` uses disk operations. It performs a lot of microscopic writes, instead of using the OS caching. I'm not convinced this can be called a bug... it's probably something which geth team will want to optimize some day.
As for now I guess I will just try to stay patient....

**Comment by karalabe** [comment]
Yes, the database trashing is a known issue. Fixing it is not so simple though as we need to completely change the database layout for a lot of things. Part of them have already been done and merged into develop quite some time ago. Part still needs to be done (if we find the time for it). You could try running an instance based on `develop` and see how that performs. (note, you probably don't want to update an existing instance to it now as it will update the database schema, making a downgrade impossible afterwards).

**Comment by kowalski** [comment]
Well, finally after 4 days I managed to sync both public and testnet chains. I have their snapshots stored on EBS and hopefully I will never have to re-do the first-sync operation again. 
If someone stumbles upon the same issue, my advice is to not use EBS for the datadir during first sync. Instead you should use the ephemeral storage (the one physically attached to the machine) and only copy the result to EBS after the first sync is finished. 
I think we can close this issue.

**Comment by mamija** [comment]
I have the same problem, unfortunately I am not a programmer,  so sy could explain how can I fully snyc my wallet

**Comment by kowalski** [comment]
@mamija the problem in my case was that I put the `.ethereum` folder on a mounted volume which had only 100 IOPS provisioned. It turns out that `geth` requires *much* more than that. 
The solution was to use faster type of storage during the first sync, than stop `geth` and move the folder to the target drive.
I doubt that you have the same problem  - if you're not a programmer I find it unlikely that you are fiddling with EBS volumes...

**Comment by shenmanny2** [comment]
I run geth in docker but sync going very slow after the block number bigger than 487215 (volume space about 5 G) on testnet

**Comment by unixtastic** [comment]
T2 type AWS instances have burstable CPU, they are not designed for consistently high CPU usage. Can you please check that you are not running out of CPU credits? You can check in the AWS console.

**Comment by GiuseppeSaponieri** [comment]
INFO [01-05|13:23:40] Imported new state entries               count=573  elapsed=2.004ms      processed=7375760 pending=9956  retry=461  duplicate=4532 unexpected=10575
WARN [01-05|13:24:17] Stalling state sync, dropping peer       peer=7148f6b8b83c2a2d
WARN [01-05|13:25:34] Stalling state sync, dropping peer       peer=68a73807eff4a2d8

Can someone tell me what's going on with this sync. Aroun 24 hours now.
